### PR TITLE
PubSub Emulator and firebase emulators init

### LINKED ISSUE #1748: FR: PubSub emulator support
I tried running cloud function locally using
Firebase emulators:start

the response i got

i  Starting emulators: ["functions","database","hosting"]
⚠  Your requested "node" version "8" doesn't match your global version "10"
✔  functions: Emulator started at http://localhost:5001
i  database: Emulator logging to database-debug.log
✔  database: Emulator started at http://localhost:9000
i  database: For testing set FIREBASE_DATABASE_EMULATOR_HOST=localhost:9000
i  hosting: Serving hosting files from: public
✔  hosting: Local server: http://localhost:5000
✔  hosting: Emulator started at http://localhost:5000
i  functions: Watching "/home/roc***/dev/vue/ev***/functions" for Cloud Functions...
i  functions[scheduledFunction]: function ignored because the pubsub.googleapis.com emulator does not exist or is not running.
✔  All emulators started, it is now safe to connect.

how do i get the pubsub emulator to work locally?

### PR DESCRIPTION
**Pull request by samtstern**
### Description

- Adds a PubSub emulator (fixes #1748)
- Adds a new subcommand command `firebase init emulators`

TODO
- [x] `init emulators`
- [x] `setup:emulators:pubsub`
- [x] Remove `--pubsub` flag
- [x] Standardize logging
- [x] Clarify interactions with `firebase.json`
- [x] Detect emulator starting from `firebase.json`
- [x] Make sure it works with `--only`
- [x] End to end testing
- [x] Move JAR to proper location
- [x] Switch to downloading ZIP

Dependencies:
- [x] gcloud approval for JAR versioning / distribution
- [x] API approval for CLI command

#### New Init

```bash
=== Project Setup

First, let's associate this project directory with a Firebase project.
You can create multiple project aliases by running firebase use --add, 
but for now we'll just set up a default project.

i  .firebaserc already has a default project, using fir-dumpster.

=== Emulators Setup
? Which Firebase emulators do you want to set up? Press Space to select emulators, then Ent
er to confirm your choices. Functions, Pubsub
? Which port do you want to use for the functions emulator? 5001
? Which port do you want to use for the pubsub emulator? 8085
? Would you like to download the emulators now? Yes
i  pubsub: downloading cloud-pubsub-emulator-v0.1.0.jar...
Progress: =================================================================> (100% of 33MB

i  Writing configuration info to firebase.json...
i  Writing project information to .firebaserc...

✔  Firebase initialization complete!
samstern@freewifi:/tmp/tmp.ITmKYw5141 
$ cat firebase.json 
{
  "emulators": {
    "functions": {
      "port": 5001
    },
    "pubsub": {
      "port": 8085
    }
  }
}
```

#### PubSub Functions

Tests use these functions:

```js
const functions = require('firebase-functions');
const admin = require('firebase-admin');
admin.initializeApp();

const { PubSub } = require('@google-cloud/pubsub');
const pubsub = new PubSub();

exports.pubsubWriter = functions.https.onRequest(async (req, res) => {
    console.log("Pubsub Emulator:", process.env.PUBSUB_EMULATOR_HOST);

    const msg = await pubsub.topic('test-topic').publishJSON({
        foo: 'bar',
        date: new Date()
    }, { attr1: 'value' });

    res.json({
        published: msg
    })
});

exports.pubsubFn = functions.pubsub.topic('test-topic').onPublish(async (msg, ctx) => {
    console.log('Received pubsub');
    console.log('json', JSON.stringify(msg.json), 'attrs', JSON.stringify(msg.attributes));
    return true;
});
```

Starting the emulators:

All emulators:
```bash
$ npx firebase emulators:start
✔  functions: Using node@10 from host.
✔  functions: Emulator started at http://localhost:5001
i  pubsub: Emulator logging to pubsub-debug.log
✔  pubsub: Emulator started at http://localhost:8085
i  functions: Watching "/tmp/tmp.e4vfaKijHH/functions" for Cloud Functions...
✔  functions[pubsubWriter]: http function initialized (http://localhost:5001/fir-dumpster/us-central1/pubsubWriter).
✔  functions[pubsubFn]: pubsub function initialized.
✔  All emulators started, it is now safe to connect
```

Only PubSub:
```bash
$ npx firebase emulators:start --only pubsub
i  pubsub: Emulator logging to pubsub-debug.log
✔  pubsub: Emulator started at http://localhost:8085
✔  All emulators started, it is now safe to connect.
```

PubSub function chain:

```bash
i  functions: Beginning execution of "pubsubWriter"
>  Pubsub Emulator: localhost:8085
i  functions: Finished "pubsubWriter" in ~1s
i  functions: Beginning execution of "pubsubFn"
>  Received pubsub
>  json {"foo":"bar","date":"2019-11-04T21:48:31.431Z"} attrs {"attr1":"value"}
i  functions: Finished "pubsubFn" in ~1s
```

### Sample Commands

```bash
$ firebase init emulators
```

### CODE REVIEW

File: src/serve/javaEmulators.ts [hunk 1]
```diff
@@ -46,18 +50,40 @@ const EmulatorDetails: { [s in JavaEmulators]: JavaEmulatorDetails } = {
     localPath: path.join(CACHE_DIR, "cloud-firestore-emulator-v1.10.0.jar"),
     namePrefix: "cloud-firestore-emulator",
   },
+
+  // TODO: This is not a real production bucket! This is just in test project. Need to move it.
```
Comment (samtstern) [inline src/serve/javaEmulators.ts]: STOPSHIP

File: .travis.yml [hunk 2]
```diff
@@ -2,6 +2,7 @@ language: node_js
 node_js:
   - "8"
   - "10"
+sudo: true
```
Comment (samtstern) [inline .travis.yml]: Not sure why Travis is failing...

File: src/emulator/types.ts [hunk 3]
```diff
@@ -57,20 +76,70 @@ export interface JavaEmulatorCommand {
...
+export interface JavaEmulatorOptions {
   cacheDir: string;
   remoteUrl: string;
   expectedSize: number;
   expectedChecksum: string;
-  localPath: string;
   namePrefix: string;
 }
 
+export abstract class JavaEmulatorDetails {
```
Comment (samtstern) [inline src/emulator/types.ts]: Previously this was just an `interface` but now that we have to support both `zip` and `jar` I found it easier to abstract this into a class where the emulators can specify multiple paths (not just the path where the download should go)

File: package.json [hunk 4]
```diff
@@ -101,6 +102,7 @@
     "tcp-port-used": "^1.0.1",
     "tmp": "0.0.33",
     "universal-analytics": "^0.4.16",
+    "unzipper": "^0.10.5",
```
Comment (bkendall) [inline package.json]: Can I suggest using [`archiver`](https://www.npmjs.com/package/archiver) which is already in `package.json`?

File: scripts/set-default-credentials.sh [hunk 5]
```diff
@@ -9,10 +9,14 @@ if [ "${TRAVIS}" != "true" ]; then
...
-if [ "${TRAVIS_REPO_SLUG}" == "firebase/firebase-tools" ]; then
-  GOOGLE_APPLICATION_CREDENTIALS="${CWD}/scripts/creds-public.json"
+if [ "${LOCAL}" != "true" ]; then
+  GOOGLE_APPLICATION_CREDENTIALS="${CWD}/scripts/creds-private.json"
+  if [ "${TRAVIS_REPO_SLUG}" == "firebase/firebase-tools" ]; then
+    GOOGLE_APPLICATION_CREDENTIALS="${CWD}/scripts/creds-public.json"
+  fi
+  export GOOGLE_APPLICATION_CREDENTIALS
+else
+  echo "Not setting GOOGLE_APPLICATION_CREDENTIALS because LOCAL=${LOCAL}"
```
Comment (bkendall) [inline scripts/set-default-credentials.sh]: Why was this necessary?

File: scripts/set-default-credentials.sh [hunk 6]
```diff
@@ -9,10 +9,14 @@ if [ "${TRAVIS}" != "true" ]; then
   export TRAVIS_REPO_SLUG="firebase/firebase-tools"
 fi
 
-GOOGLE_APPLICATION_CREDENTIALS="${CWD}/scripts/creds-private.json"
-if [ "${TRAVIS_REPO_SLUG}" == "firebase/firebase-tools" ]; then
-  GOOGLE_APPLICATION_CREDENTIALS="${CWD}/scripts/creds-public.json"
+if [ "${LOCAL}" != "true" ]; then
```
Comment (bkendall) [inline scripts/set-default-credentials.sh]: Might I suggest

```bash
if [ -z "$LOCAL" ]
```

https://stackoverflow.com/questions/3601515/how-to-check-if-a-variable-is-set-in-bash

File: scripts/triggers-end-to-end-tests/run.spec.js [hunk 7]
```diff
@@ -335,3 +339,61 @@ describe("database and firestore emulator function triggers", function() {
...
+  });
+
+  after(function(done) {
+    this.timeout(EMULATORS_SHUTDOWN_DELAY_MS);
+    if (test) {
+      test.stopEmulators(done);
+      return;
+    }
+    done();
+  });
```
Comment (bkendall) [inline scripts/triggers-end-to-end-tests/run.spec.js]: Typically this goes after the `before` block, before the tests (`it`s). It's considered setup

File: src/commands/setup-emulators-pubsub.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import * as Command from "../command";
+
+const downloadEmulator = require("../emulator/download");
+const emulatorName = "pubsub";
```
Comment (bkendall) [inline src/commands/setup-emulators-pubsub.ts]: s/emulatorName/EMULATOR_NAME (constant case)

and this declaration should be its own block

File: src/commands/setup-emulators-pubsub.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+
+module.exports = new Command("setup:emulators:" + emulatorName)
+  .description("downloads the " + emulatorName + " emulator")
```
Comment (bkendall) [inline src/commands/setup-emulators-pubsub.ts]: prefer template strings

File: src/emulator/controller.ts [hunk 10]
```diff
@@ -76,8 +76,22 @@ export async function cleanShutdown(): Promise<boolean> {
...
+  } else if (options.except) {
+    targets = _.difference(targets, options.except.split(","));
+  }
+
+  return targets;
+}
+
 export function shouldStart(options: any, name: Emulators): boolean {
-  const targets: string[] = filterTargets(options, VALID_EMULATOR_STRINGS);
+  const targets: string[] = filterEmulatorTargets(options);
```
Comment (bkendall) [inline src/emulator/controller.ts]: this typing (`string[]`) is redundant

File: src/emulator/controller.ts [hunk 11]
```diff
@@ -92,14 +106,13 @@ export async function startAll(options: any): Promise<void> {
   // }
   //
   // The list of emulators to start is filtered two ways:
-  // 1) The service must have a top-level entry in firebase.json
+  // 1) The service must have a top-level entry in firebase.json or an entry in the emulators{} object
   // 2) If the --only flag is passed, then this list is the intersection
-  const targets: string[] = filterTargets(options, VALID_EMULATOR_STRINGS);
+  const targets: string[] = filterEmulatorTargets(options);
```
Comment (bkendall) [inline src/emulator/controller.ts]: again, unnecessary type. I see a bunch coming, I'm just not going to worry about them any more

File: src/emulator/controller.ts [hunk 12]
```diff
@@ -221,6 +234,23 @@ export async function startAll(options: any): Promise<void> {
     await startEmulator(hostingEmulator);
   }
 
+  if (shouldStart(options, Emulators.PUBSUB)) {
+    if (!projectId) {
+      throw new FirebaseError(
+        "Can't start pubsub emulator without projectId, run 'firebase init' or pass the --project flag"
```
Comment (bkendall) [inline src/emulator/controller.ts]:
```suggestion
        "Cannot start the PubSub emulator without a project; run 'firebase init' or provide the --project flag"
```

File: src/emulator/download.ts [hunk 13]
```diff
@@ -9,41 +9,57 @@ import * as javaEmulators from "../serve/javaEmulators";
 import * as tmp from "tmp";
 import * as fs from "fs-extra";
 import * as path from "path";
+import * as unzipper from "unzipper";
```
Comment (bkendall) [inline src/emulator/download.ts]: ah, here's the usage. Maybe prefer `archiver`?

File: src/emulator/functionsEmulator.ts [hunk 14]
```diff
@@ -524,6 +529,35 @@ export class FunctionsEmulator implements EmulatorInstance {
     });
   }
 
+  addPubsubTrigger(projectId: string, definition: EmulatedTriggerDefinition): Promise<boolean> {
```
Comment (bkendall) [inline src/emulator/functionsEmulator.ts]: use async?

```typescript
async addPubsubTrigger(...
  if () { return false; }
  pubsubEmulator.addTrigger()
  return true;
}
```

File: src/emulator/pubsubEmulator.ts [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+
+    EmulatorLogger.logLabeled(
+      "DEBUG",
+      "pubsub",
+      `Executing ${topicTriggers.size} matching triggers`
+    );
+    for (const trigger of topicTriggers) {
+      const body = {
+        context: {
+          // TODO(samstern): Is this an acceptable eventId?
```
Comment (bkendall) [inline src/emulator/pubsubEmulator.ts]: Is it? :)

File: src/emulator/types.ts [hunk 16]
```diff
@@ -57,20 +76,70 @@ export interface JavaEmulatorCommand {
...
+  get downloadPath(): string {
+    return this.localPath;
+  }
+
+  get binaryPath(): string {
+    return this.localPath;
+  }
+
+  get unzipDir(): string | undefined {
+    return undefined;
```
Comment (bkendall) [inline src/emulator/types.ts]: Should this just throw instead of being `| undefined`?

File: src/emulator/types.ts [hunk 17]
```diff
@@ -57,20 +76,70 @@ export interface JavaEmulatorCommand {
...
+  }
+}
+
+export class ZipEmulatorDetails extends JavaEmulatorDetails {
+  constructor(
+    name: string,
+    private localPath: string,
+    private _unzipDir: string,
```
Comment (bkendall) [inline src/emulator/types.ts]: I cringe when I see underscore-prefixed variables. Can this just be `internalUnzipDir` or something? see the TS style guide for more... Also, it has discussions about getters that's interesting (go/ts-style)

File: src/emulator/types.ts [hunk 18]
```diff
@@ -1,21 +1,40 @@
 import { ChildProcess } from "child_process";
 import { EventEmitter } from "events";
+import * as path from "path";
 
 export enum Emulators {
```
Comment (bkendall) [inline src/emulator/types.ts]: I think enums are usually better off defined in the singular, but...

File: src/init/features/emulators.ts [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+          type: "input",
+          name: "port",
+          message: `Which port do you want to use for the ${clc.underline(selected)} emulator?`,
+          default: Constants.getDefaultPort(selected as Emulators),
+        },
+      ]);
+    }
+  }
+
+  if (selections.emulators.length > 0) {
```
Comment (bkendall) [inline src/init/features/emulators.ts]: `.length` is sufficient

File: src/init/features/emulators.ts [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+      },
+    ]);
+  }
+
+  if (selections.download) {
+    for (const selected of selections.emulators) {
+      if (isJavaEmulator(selected)) {
+        await downloadIfNecessary(selected);
+      }
+    }
```
Comment (bkendall) [inline src/init/features/emulators.ts]: Can/should this be done in parallel?

File: src/serve/javaEmulators.ts [hunk 21]
```diff
@@ -88,13 +115,24 @@ function _getCommand(emulator: JavaEmulators, args: { [s: string]: any }): JavaE
     const argKey = "--" + key;
     const argVal = args[key];
 
-    cmdLineArgs.push(argKey, argVal);
+    if (!argVal || argVal === "") {
```
Comment (bkendall) [inline src/serve/javaEmulators.ts]: `!argVal` is sufficient, right?

File: src/serve/javaEmulators.ts [hunk 22]
```diff
@@ -194,17 +232,26 @@ export async function stop(targetName: JavaEmulators): Promise<void> {
   });
 }
 
+export function downloadIfNecessary(targetName: JavaEmulators) {
```
Comment (bkendall) [inline src/serve/javaEmulators.ts]: `export async function` that returns `Promise<void>`?

File: package.json (hunk 4 above)
Comment (samtstern) [reply package.json]: Apparently that library only zips, does not unzip:
https://github.com/archiverjs/node-archiver/issues/97

File: scripts/set-default-credentials.sh (hunk 5 above)
Comment (samtstern) [reply scripts/set-default-credentials.sh]: Makes it much easier to run these integration tests locally and re-use whatever default credentials the CLI already has.

File: src/commands/setup-emulators-pubsub.ts (hunk 8 above)
Comment (samtstern) [reply src/commands/setup-emulators-pubsub.ts]: :+1: sorry about all the style in this file I just copied the `setup-database-emulator` file and didn't think about it.

File: src/emulator/types.ts (hunk 16 above)
Comment (samtstern) [reply src/emulator/types.ts]: No we do:
```
if (emulator.unzipDir) {
    await unzip(emulator.downloadPath, emulator.unzipDir);
}
```

File: src/emulator/types.ts (hunk 18 above)
Comment (samtstern) [reply src/emulator/types.ts]: Yeah I regret this from the start but don't want to bloat the diff changing it now.

File: src/init/features/emulators.ts (hunk 20 above)
Comment (samtstern) [reply src/init/features/emulators.ts]: No because there is a progressbar with each download.

File: src/serve/javaEmulators.ts (hunk 22 above)
Comment (samtstern) [reply src/serve/javaEmulators.ts]: I think declaring a function `async` without using `await` inside it is considered rude.

Comment (bkendall) [reply src/serve/javaEmulators.ts]: The return type of this function is `Promise<something> | undefined`. If it's async, you could reduce it to `Promise<any>`. For that matter, do you even care what it returns? Maybe the `if (!hasEmulator)` should just await `downloadEmulator` and then the return type is `Promise<void>`. Thoughts?

Comment (samtstern) [reply src/serve/javaEmulators.ts]: I think you're looking at an outdated diff, the function is now:
```ts
export function downloadIfNecessary(targetName: JavaEmulators): Promise<any> {
  const emulator = EmulatorDetails[targetName];
  const hasEmulator = fs.existsSync(emulator.binaryPath);

  if (!hasEmulator) {
    return downloadEmulator(targetName);
  } else {
    return Promise.resolve();
  }
}
```

Comment (samtstern) [reply src/serve/javaEmulators.ts]: But anyway I can move it to `async` and await the inner ca..

File: scripts/set-default-credentials.sh [hunk 23]
```diff
@@ -9,10 +9,14 @@ if [ "${TRAVIS}" != "true" ]; then
   export TRAVIS_REPO_SLUG="firebase/firebase-tools"
 fi
 
-GOOGLE_APPLICATION_CREDENTIALS="${CWD}/scripts/creds-private.json"
-if [ "${TRAVIS_REPO_SLUG}" == "firebase/firebase-tools" ]; then
-  GOOGLE_APPLICATION_CREDENTIALS="${CWD}/scripts/creds-public.json"
+if [ -z "${LOCAL}" ]; then
```
Comment (ryanpbrewster) [inline scripts/set-default-credentials.sh]: Since we're using bash we can do
```
if [[ -z $LOCAL ]]; then
```

The [double-brackets](https://stackoverflow.com/questions/669452/is-double-square-brackets-preferable-over-single-square-brackets-in-ba) are slightly more ergonomic to use.

File: src/serve/javaEmulators.ts [hunk 24]
```diff
@@ -194,17 +232,28 @@ export async function stop(targetName: JavaEmulators): Promise<void> {
   });
 }
 
+export async function downloadIfNecessary(targetName: JavaEmulators): Promise<any> {
```
Comment (ryanpbrewster) [inline src/serve/javaEmulators.ts]: should this be `Promise<void>`?

File: src/serve/javaEmulators.ts [hunk 25]
```diff
@@ -88,13 +115,24 @@ function _getCommand(emulator: JavaEmulators, args: { [s: string]: any }): JavaE
     const argKey = "--" + key;
     const argVal = args[key];
 
-    cmdLineArgs.push(argKey, argVal);
+    if (!argVal) {
```
Comment (ryanpbrewster) [inline src/serve/javaEmulators.ts]: how likely is this to misbehave if I pass in a boolean like `--fork=false`?

File: src/emulator/types.ts (hunk 3 above)
Comment (ryanpbrewster) [reply src/emulator/types.ts]: Can we hide that as an implementation detail of `download`? These two classes look like they share a lot of logic.

File: src/emulator/types.ts [hunk 26]
```diff
@@ -1,21 +1,40 @@
...
   Emulators.FIRESTORE,
   Emulators.DATABASE,
   Emulators.HOSTING,
+  Emulators.PUBSUB,
 ];
 
+export const JAVA_EMULATORS = [Emulators.FIRESTORE, Emulators.DATABASE, Emulators.PUBSUB];
+
+export function isJavaEmulator(value: string): value is JavaEmulators {
+  if (isEmulator(value)) {
```
Comment (ryanpbrewster) [inline src/emulator/types.ts]: equivalent to
```
return isEmulator(value) && JAVA_EMULATORS.indexOf(value) >= 0;
```

also, do we need the first clause?

File: src/emulator/pubsubEmulator.ts [hunk 27]
```diff
@@ -0,0 +1,185 @@
...
+        },
+      };
+
+      const functionsUrl = `http://localhost:${functionsPort}/functions/projects/${
+        this.args.projectId
+      }/triggers/${trigger}`;
+      request.post(
+        functionsUrl,
+        {
+          body: JSON.stringify(body),
```
Comment (ryanpbrewster) [inline src/emulator/pubsubEmulator.ts]: I think this can be
```
{ body, json: true }
```
which should additionally set the content-type for you

File: src/serve/javaEmulators.ts (hunk 25 above)
Comment (samtstern) [reply src/serve/javaEmulators.ts]: Very ... good call.

File: src/emulator/types.ts (hunk 26 above)
Comment (samtstern) [reply src/emulator/types.ts]: Thanks.  And yeah we do, to make TS not complain.  It makes `value` definitely an `Emulators`

File: src/emulator/types.ts (hunk 3 above)
Comment (samtstern) [reply src/emulator/types.ts]: Ok you're right this sucked.  I changed it back to an interface but now I separated download constants from things like `instance` and `stdout` since they never really belonged together:
https://github.com/firebase/firebase-tools/pull/1755/commits/c58cdc6669a71e0b0b84b673cc9b5453895b84c8

File: src/emulator/controller.ts [hunk 28]
```diff
@@ -75,8 +75,22 @@ export async function cleanShutdown(): Promise<boolean> {
...
+export function filterEmulatorTargets(options: any): string[] {
+  let targets = VALID_EMULATOR_STRINGS.filter((e) => {
+    return options.config.has(e) || options.config.has(`emulators.${e}`);
+  });
+
+  if (options.only) {
+    targets = _.intersection(targets, options.only.split(","));
+  } else if (options.except) {
+    targets = _.difference(targets, options.except.split(","));
+  }
```
Comment (ryanpbrewster) [inline src/emulator/controller.ts]: should we throw an error if someone tries `--only database --except firestore`? right now it seems like we just ignore the `--except` if you provide both

File: src/emulator/pubsubEmulator.ts [hunk 29]
```diff
@@ -0,0 +1,186 @@
...
+    this.triggers = new Map();
+    this.subscriptions = new Map();
+  }
+
+  async start(): Promise<void> {
+    return javaEmulators.start(Emulators.PUBSUB, this.args);
+  }
+
+  async connect(): Promise<void> {
+    return Promise.resolve();
```
Comment (ryanpbrewster) [inline src/emulator/pubsubEmulator.ts]: I think this can be
```
async connect(): Promise<void> {}
```

File: src/emulator/pubsubEmulator.ts [hunk 30]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  async stop(): Promise<void> {
+    return javaEmulators.stop(Emulators.PUBSUB);
```
Comment (ryanpbrewster) [inline src/emulator/pubsubEmulator.ts]: `await ...`

File: src/emulator/pubsubEmulator.ts [hunk 31]
```diff
@@ -0,0 +1,186 @@
...
+          "pubsub",
+          `Functions emulator response: HTTP ${res.statusCode} ${JSON.stringify(res.body)}`
+        );
+      }
+
+      remaining--;
+      if (remaining <= 0) {
+        EmulatorLogger.logLabeled("DEBUG", "pubsub", `Acking message ${message.id}`);
+        message.ack();
+      }
```
Comment (ryanpbrewster) [inline src/emulator/pubsubEmulator.ts]: I don't understand what this code is doing. Can we add a comment explaining the workflow here?

File: src/serve/javaEmulators.ts [hunk 32]
```diff
@@ -26,68 +25,92 @@ const EMULATOR_INSTANCE_KILL_TIMEOUT = 2000; /* ms */
...
+    binary: getExecPath(Emulators.PUBSUB)!,
     args: [],
     optionalArgs: ["port", "host"],
     joinArgs: true,
   },
 };
 
+function getExecPath(name: JavaEmulators): string {
+  const details = getDownloadDetails(name);
+  return details.binaryPath || details.downloadPath;
```
[the remark this answers is not in the document]
Comment (samtstern) [reply src/serve/javaEmulators.ts]: No we are not.  Sadness.

File: src/emulator/controller.ts (hunk 28 above)
Comment (samtstern) [reply src/emulator/controller.ts]: Fun fact: `emulators:start` and `exec` don't even accept `--except` for now so I will just drop this branch.

File: src/emulator/pubsubEmulator.ts (hunk 29 above)
Comment (samtstern) [reply src/emulator/pubsubEmulator.ts]: Unfortunately:
```
  89:34  error  Unexpected empty async method 'connect'
```

**Comment by ultraGentle** [comment]
Hello, beginner Firebase user here: 

Question: Manual triggering of scheduled functions <-- is this supported by the emulator?  

(I came here via #1748, where that was requested.)

**Comment by samtstern** [comment]
That's not yet supported but it's something we're working on!
